cpp/src/arrow/compute/kernels/take_internal.h

```diff
@@ -583,10 +583,8 @@ class TakerImpl<IndexSequence, UnionType> : public Taker<IndexSequence> {
       indices.set_never_out_of_bounds();
 
       // Allocate temporary storage for the offsets of all valid slots
-      // NB: Overestimates required space when indices and union_array are
-      //     not null at identical positions.
       auto child_offsets_storage_size =
-          indices.length() - std::max(union_array.null_count(), indices.null_count());
+          std::accumulate(child_counts.begin(), child_counts.end(), 0);
```

Comment:
@bkietz This code could lead to negative buffer size. I think the new version is accurate. Can you confirm?

Reply:
@pitrou looks good to me. Easier to understand, too. +1